trieste/space.py

```diff
@@ -329,3 +346,124 @@ def __mul__(self, other: Box) -> Box:
+        return self._tags
+
+    @property
+    def dimension(self) -> TensorType:
+        """The input dimensionality of this product search space."""
+        return self._dimension
+
+    def get_subspace(self, tag: str) -> SearchSpace:
+        """Return the domain of a particular subspace."""
+        return self._spaces[tag]
+
+    def get_subspace_component(self, tag: str, values: TensorType) -> TensorType:
+        """
+        Returns the components of ``values`` lying in a particular subspace.
+
+        :param value: Points from the :class:`DecomposableSearchSpace` of shape [N,D].
+        :return: The context components of ``values` of shape [M, D], where
+            M is the dimensionality of the specified subspace.
+        """
+
+        starting_index_of_subspace = self._subspace_starting_indicies[tag]
+        ending_index_of_subspace = starting_index_of_subspace + self._subspace_sizes[tag]
+        return values[:, starting_index_of_subspace:ending_index_of_subspace]
+
+    def __contains__(self, value: TensorType) -> bool | TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point is a
+        member if each of its subspace components lie in each subspace.
+
+        Recall that individual points in product spaces are represented with their inputs in the
+        same order as specified when initializing the space.
+
+        :param value: A point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
+            scalar boolean `TensorType` instead of the `bool` itself.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``value`` has a different
+            dimensionality from the search space.
+        """
+
+        tf.debugging.assert_equal(
+            tf.shape(value),
+            self.dimension,
+            message="value must have same dimensionality as search space",
+        )
+        value = tf.expand_dims(value, 0)
+        in_each_subspace = [
+            self._spaces[tag].__contains__(self.get_subspace_component(tag, value)[0, :])
+            for tag in self._tags
+        ]
+        return tf.math.reduce_all(in_each_subspace)
+
+    def sample(self, num_samples: int) -> TensorType:
+        """
+        Sample randomly from the space by sampling from each subspace
+        and concatenating the resulting samples.
+
+        :param num_samples: The number of points to sample from this search space.
+        :return: ``num_samples`` i.i.d. random points, sampled uniformly, and without replacement,
+            from this search space with shape '[num_samples, D]' , where D is the search space
+            dimension.
+        """
+        tf.debugging.assert_non_negative(num_samples)
+
+        subspace_samples = [self._spaces[tag].sample(num_samples) for tag in self._tags]
+        return tf.concat(subspace_samples, -1)
```

Comment:
its important that the sample method of this class just returns a tensor, as this will be used when maximizing acquisition functions over the product space

Reply:
There is no leeway here, clients of this class will just call `space.sample()` and expect the same result no matter which space implementation it is.